garage/misc/git_handler.py

```diff
@@ -0,0 +1,382 @@
+                                      self.branch_name, self.restore_sha)
+
+    def _verify_target(self, target):
+        """Verify the key-value target is valid.
+
+        Parameters
+        ----------
+        - target: same description as in method run_experiment_on_target.
+
+        Returns
+        -------
+        If the target is valid, the corresponding SHA.
+
+        """
+        target = [target.strip() for target in target.split(":")]
+        sha = ""
+        assert (len(target) == 2 and target[0]
+                and target[1]), self._get_target_use_message()
+        try:
+            verify = self.valid_targets[target[0]]
+            sha = verify(target[1])
+        except KeyError:
+            print(self._get_target_use_message())
+            raise
+        return sha
+
+    def _get_target_use_message(self):
+        """Print a use message if the user passed an invalid target."""
+        return colored(
+            "\nInvalid target: targets can be defined with a SHA, branch " \
+            "name (local or remote) or tag\nas shown " \
+            "in the following examples:\nsha: 8d54s23a" \
+            "\nbranch: my_branch\nbranch: remote/master\ntag: my_tag\n",
+            "yellow")
+
+    def _switch_to_target(self, target_branch_name, target_sha):
+        """Clean working directory and check out the target branch.
+
+        The target branch is created during the checkout operation.
+        SIGINT is temporarily blocked while performing the git operations to
+        avoid corrupting the working directory.
+        """
+        # Block SIGINT to avoid ending in an unknown state
+        signal.pthread_sigmask(signal.SIG_BLOCK, {signal.SIGINT})
+        # Clean working directory
+        self.reset()
+        # Change to target directory using a new branch
+        self.checkout_new_branch(target_branch_name, target_sha)
+        signal.pthread_sigmask(signal.SIG_UNBLOCK, {signal.SIGINT})
+
+    def _run_experiment(self, exp_args):
+        """Call method run_experiment at garage.misc.instrument.
+
+        The git target is removed from the dictionary of arguments passed to
+        run_experiment, and the key arguments are updated directly into the
+        run_experiment arguments.
+
+        Parameters
+        ----------
+        - exp_args: it's a dictionary of arguments obtained from the function
+            run_experiment at garage.misc.instrument, that will be passed to
+            the experiment once the target is checked out.
+
+        """
+        # Imports are here to avoid circular dependencies within garage.misc
+        import garage.misc.instrument
+        from garage.misc import instrument
+        if "git_target" in exp_args:
+            exp_args.pop("git_target")
+        if "kwargs" in exp_args:
+            kwargs = exp_args["kwargs"]
+            exp_args.pop("kwargs")
+            exp_args.update(kwargs)
+        importlib.reload(garage.misc.instrument)
```

Comment:
For the cases where the target commit to test also includes changes in the function run_experiment (which is in garage.misc.instrument). Since the module is loaded in cache before performing the switching, the function run_experiment won't be that of the target commit unless it's reloaded.